### DISCUSSION TITLE
Extension resolution for data provider

### DISCUSSION THREAD

**Discussion by jedel1043** [unicode-org/icu4x#2429] [Q&A] [answered]
Is there an API to verify if an extension value for a specific `LanguageIdentifier` in a provider is supported, or plans to provide such API?

Extended:

ECMA-402 defines [`ResolveLocale`](https://tc39.es/ecma402/#sec-resolvelocale), which goes through the process of checking if a language is supported by the provider (solved in #2374), and choosing a valid set of unicode extensions supported by the provider for the selected language.

This process is aid by the [`[[LocaleData]]`](https://tc39.es/ecma402/#sec-internal-slots) internal slot, which has all the information of every supported unicode extension for every "supported" language. However, https://github.com/unicode-org/icu4x/discussions/2374#discussioncomment-3393598 clarifies that providers don't have a bounded list of supported languages because of locale fallback, so I doubt providers would also contain an equivalent to `[[LocaleData]]`.

From some tests I did, I can apparently do data requests like on #2374 to check if an extension is supported by the provider. However, I don't know if this is a reliable method to do this. I also would need to consider default values for every extension, but said values could differ between languages. (would #1833 solve that?) Are there any APIs I'm missing that could be used in this case?

**Comment by sffc** [comment, marked as answer]
Clients are intended to treat locale fallback like a black box. As long as you are using a properly configured LocaleFallbackProvider in ICU4X, all supported extension keywords will be resolved automatically.

Note that Locale and DataLocale support extension keywords but not LanguageIdentifier.

**Reply by jedel1043** [reply]
Ah, so locales with fallback automatically return the default value of an extension if the provider doesn't support the requested value?

**Reply by sffc** [reply]
Correct

See Thai DateTimeFormatter for an example. It defaults to the Buddhist calendar if the -u-ca extension is missing.

**Reply by jedel1043** [reply]
Nice! That simplifies the logic by a lot, since I just need to pass the locale directly to the component I want to use, then recover the resolved settings and locale from the created component, right?

Though, `hc` is an exception, because ECMA-402 defines `hour12` which forcibly converts the default value to an 11/12 hour format, meaning I'd need to manually fetch the `TimeLengthV1` data to get the default time before constructing the `DateTimeFormatter`. I'm hoping this is the only exception I'd need to handle 😅
